### PR TITLE
Audit queues

### PR DESCRIPTION
**Pull request by nwoodHO**

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java [hunk 1]
```diff
@@ -2,27 +2,98 @@
...
     }
 
-    public AuditData createAudit(String correlationID, String raisingService, String before, String after, String namespace, String type, String userID){
-        log.debug("Creating Audit: {}, ID: {}, Raised by: {}, By user: {}", correlationID, raisingService, userID);
-        AuditData auditData = new AuditData(correlationID, raisingService, before, after, namespace, type, userID);
+    public AuditData createAudit(CreateAuditDto createAuditDto) {
+        AuditData auditData = AuditData.fromDto(createAuditDto);
         auditRepository.save(auditData);
-        log.info("Creating Audit: {}, ID: {}, Raised by: {}, By user: {}", correlationID, raisingService, userID);
+        log.info("Created Audit: UUID: {}, Correlation ID: {}, Raised by: {}, Payload: {}, By user: {}, at timestamp: {}",
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java]: Don't log out the payload

File: src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java [hunk 2]
```diff
@@ -2,27 +2,98 @@
...
+    }
+
+    public List<AuditData> getAuditDataByUserIDByDateRange(String userID, String fromDate, String toDate, int page, int limit){
+        log.info("Requesting audits for User IDL {}, from dates: {} to {} ", userID, fromDate, toDate);
+        pageRequest = PageRequest.of(page,limit);
+        return auditRepository.findAuditDataByUserIDAndDateRange(userID, convertLocalDateToStartOfLocalDateTime(fromDate), convertLocalDateToEndOfLocalDateTime(toDate), pageRequest);
+    }
+
+
+    public LocalDateTime convertLocalDateToStartOfLocalDateTime(String date){
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java]: could these be private

File: src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java [hunk 3]
```diff
@@ -2,27 +2,98 @@
...
+        return auditRepository.findAuditData(lastWeek, pageRequest);
+    }
+
+    public List<AuditData> getAuditDataByDateRange(String fromDate, String toDate, int page, int limit){
+        log.info("Requesting all audits for dates: {} to {} ", fromDate, toDate);
+        pageRequest = PageRequest.of(page,limit);
+        return auditRepository.findAuditDataByDateRange(convertLocalDateToStartOfLocalDateTime(fromDate), convertLocalDateToEndOfLocalDateTime(toDate), pageRequest);
+    }
+
+    public List<AuditData> getAuditDataByCorrelationID(String correlationID, int page, int limit){
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/audit/AuditDataService.java]: for correlation ID you shouldn't need the date range filter. Correlation IDs relate to a single transaction so will only span a few minutes at most.

File: src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/model/AuditData.java [hunk 4]
```diff
@@ -56,18 +62,82 @@
...
+        return new AuditData(createAuditDto.getCorrelationID(),
+                createAuditDto.getRaisingService(),
+                auditPayload,
+                createAuditDto.getNamespace(),
+                createAuditDto.getAuditTimestamp(),
+                createAuditDto.getType(),
+                createAuditDto.getUserID());
+    }
+
+    private static void validateNotNull(CreateAuditDto createAuditDto) {
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/model/AuditData.java]: I think this should be on the CreateAuditDto class or within the AuditService.

File: src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/model/AuditData.java [hunk 5]
```diff
@@ -56,18 +62,82 @@
...
+                    raisingService,
+                    createAuditDto.getAuditPayload(),
+                    namespace,
+                    auditTimestamp,
+                    type,
+                    userID);
+        }
+    }
+
+    private static String validatePayload(CreateAuditDto createAuditDto) {
```
Comment (SWhitlam) [inline src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/model/AuditData.java]: As above, I think this should be in the DTO or the Audit Service

File: build.gradle [hunk 6]
```diff
@@ -52,6 +52,8 @@ dependencies {
 	compile group: 'org.apache.httpcomponents', name: 'httpmime', version: '4.5.6'
 	compile group: 'com.amazonaws', name: 'aws-java-sdk', version: '1.11.380'
 
+    compile "com.fasterxml.jackson:jackson-bom:2.9.0"
```
Comment (nickmcmahonHOD) [inline build.gradle]: What are you using this for? no other services need to use it AFAIK.

File: build.gradle [hunk 7]
```diff
@@ -62,4 +64,10 @@ dependencies {
 	testCompile group: 'org.apache.camel', name: 'camel-test-spring', version: camelVersion
 	testCompile('org.springframework.boot:spring-boot-starter-test')
 	testCompile('org.assertj:assertj-core')
+	testCompile "com.github.tomakehurst:wiremock-standalone:2.18.0"
+
+	compile "com.google.code.gson:gson:2.8.5"
+	compile "org.json:json:20090211"
```
Comment (nickmcmahonHOD) [inline build.gradle]: What are you using this for ? (x2) no other services need to use it AFAIK.

File: src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/repository/AuditRepository.java [hunk 8]
```diff
@@ -1,15 +1,38 @@
...
 
-    AuditData findByUuid(UUID uuid);
 
+    AuditData findAuditDataByUuid(UUID uuid);
+
+    @Query(value = "SELECT a.* FROM audit_data a WHERE a.correlation_id = ?1 AND a.audit_timestamp > ?2", nativeQuery = true)
+    List<AuditData> findAuditDataByCorrelationID(String correlationID, LocalDateTime dateTime, Pageable pageRequest);
+
+    @Query(value = "SELECT a.* FROM audit_data a WHERE a.correlation_id = ?1 AND a.audit_timestamp > ?2 AND a.audit_timestamp < ?3", nativeQuery = true)
+    List<AuditData> findAuditDataByCorrelationIDAndDateRange(String correlationID, LocalDateTime dateFrom, LocalDateTime dateTo, Pageable pageRequest);
```
Comment (nickmcmahonHOD) [inline src/main/java/uk/gov/digital/ho/hocs/audit/auditdetails/repository/AuditRepository.java]: I would always use the BETWEEN syntax here for comparing the timestamps, but that's preference.

File: build.gradle (hunk 7 above)
Comment (nwoodHO) [reply build.gradle]: I removed all imports except the gson, which I use to validate the json blob.

**Pull request merged**